Clarify and extend CONTRIBUTING.md (#3569)
subspace/subspace: CONTRIBUTING.md

contributors is much larger than number of maintainers, and the wider community of users is even bigger than that.

As such, the process of contributing to Autonomys Network needs to be optimized in such a way that it doesn't overwhelm
maintainers with unnecessary notifications, or questions that can be answered by wider community. Following the practices
defined in this document below, such that contributions can be processed in the most productive way, saves time and
makes for a better experience overall.

## Bugs, improvements and feature requests

These should be submitted in such a way that they are actionable by developers. Again, if you're unsure or just have a
question - use the forum, a GitHub issue can be created later if it is necessary.

Always search for similar topics already created on the forum or here on GitHub, before creating a new one. But don't
hijack threads that look somewhat similar, they are not necessarily the same, and might not have the same root cause.

### Bugs

For bug report to be actionable it should ideally follow the following template:

<details>

> <summary>Click to see the template</summary>

Author: In rendered markdown the default is "Details" and a small triangle, which isn't clear in the middle of a long document.

instead. The `.expect()` message must convince the reviewer that it will never panic, or why panicking might be the
preferred outcome (this is rare). `.expect()` correctness should be derivable from local context if it is a standalone
function, or invariants of a data structure for an internal method. If it can't be guaranteed then `Result<T, E>` must be
used instead. There are rare exceptions to this such as block numbers that never exceed `u32` in Subspace.

Indexing into slices and similar data structures with `[]` should be avoided and replaced with `.get()` or similar
methods with explicit handling of cases where element doesn't exist.

Same goes about mathematical operations, especially in runtime logic, in most cases use explicit checked, wrapping or
saturating math, but don't use saturating math just to make things compile if business logic doesn't expect overflow to
ever happen, use checked math with `.expect()` in such cases or return an error. Avoid using `as` to convert, it can
silently truncate numbers. Prefer `into`, or use `try_into` and return the error (or `expect` if the conversion never

> fails).

Author: This is new but is common practice in our code and reviews.

and rebasing during development, reorder commits if it makes more sense that way, there is great tooling out there for
these kinds of modifications.

Reviewers should be able to go through the commits one by one and understand what is being changed, with non-trivial
changes it is often very hard to review the final diff, so going through commits helps a lot. Commits should be
well-structured and there should be a reasonable number of them. 70 commits for 100 lines or changes is most likely bad,
same with one commit that changes many thousands of lines of code in various places at once (lock files are a frequent
and notable exception). Another exception is automated changes. If you do a mass rename, format, or `cargo clippy --fix`,

> put the command in the commit message, so the reviewer can run it and check it produces the same diff.

Author: Another new thing but it's very helpful for mass automatic changes.

As explained in [commits](#commits) section above, history should be clean and readable, ready for a reviewer to go
through and make sense of the changes proposed.

Make sure to test your own code locally and/or in CI before submitting a pull request to avoid numerous updates
afterwards to fix CI (remember, those changes will trigger extra notifications that could be avoided by following this
simple rule). Make sure to add new test cases if applicable.

Pull requests should be accompanied by a reasonable description. If the change is a trivial typo, the description can be
empty, but if changes are non-trivial, make sure to explain what the changes do, why, what were the alternatives and so on,
giving the reviewer proper context into the proposed changes. If there is a relevant issue or another pull request then
link it, if there is an issue that is resolved by proposed changes or PR that becomes obsolete, use
[GitHub linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Pull requests should be ready for review, avoid opening drafts that are frequently updated, in most cases you can share WIP

> changes with someone just by sharing a branch diff. Drafts have a narrow and rare use case, use them as last resort,

Author: ```suggestion
changes with someone just by sharing a branch diff, or a PR in your own fork. Drafts have a narrow and rare use case, use them as last resort,
```

otherwise they become a huge source of distraction due to number of notifications they generate. In most cases you
either want changes to be reviewed and merged, in which case a regular pull request should be created, or you don't want

> a pull request at all. If an open pull request stops being ready for review or merging, close it or put it in draft.
> Pull requests should not stay in draft longer than a week, they can be closed now, and re-opened later when ready.

Author: ```suggestion
a pull request at all. If an open pull request stops being ready for review or merging, close the PR, and open it
when it is ready for review again.
```

Reviewer: yeap 👍🏼

> Pull requests should not stay in draft longer than a week, they can be closed now, and re-opened later when ready.

Author: This is something @vedhavyas and I talked about recently.

Reviewer: > If an open pull request stops being ready for review or merging, close it or put it in draft.

I would rather close it and open up when its ready for review. 
If direction or guidence is required, open PR in your fork and ask for clarification

Author: I split those changes into two different parts of the paragraph here:
https://github.com/autonomys/subspace/pull/3569#pullrequestreview-2911512855
